app/tournament/tournament.js

```diff
@@ -0,0 +1,101 @@
+import * as player from "./player";
+import * as match from "./match";
+import * as utils from "../common/utils";
+
+export class Tournament {
+    constructor() {
+        this.players = [];
+        this.matches = [];
+    }
+
+    addPlayer(player) {
+        if (this.existsAlready(player)) {
+            throw new TournamentError(`Duplicate player: "${player.name}".`);
+        }
+        this.players.push(player);
+    }
+
+    existsAlready(player) {
+        for (var p of this.players) {
+            if (p.name.toLowerCase() === player.name.toLowerCase()) {
+                return true;
+            }
+        };
+        return false;
+    }
+
+    generate() {
+        this.checkNumberOfPlayers();
+        this.generateEmptyMatches();
+        this.generateTournamentBracket();
+        this.pairUpPlayers();
+    }
+
+    checkNumberOfPlayers() {
+        if (!this.enaughPlayers()) {
+            throw new TournamentError(`Too few players: ${this.players.length}`);
+        }
+        if (!this.hasProperNumberOfPlayers()){
+            this.fillUpWithGhosts();
+        }
+    }
+
+    enaughPlayers() {
+        return this.players.length >= 2;
+    }
+
+    hasProperNumberOfPlayers() {
+        return Number.isInteger(Math.log2(this.players.length));
+    }
+
+    fillUpWithGhosts() {
+        var goodNumber = parseInt(Math.pow(10, this.players.length.toString(2).length), 2);
+        var i = 1;
+        while (this.players.length < goodNumber) {
+            this.addPlayer(new player.Player(`Ghost${i++}`, 1));
+        }
+    }
+
+    generateEmptyMatches() {
+        for (var i=0; i<this.totalMatches(); i++) {
+            this.matches.push(new match.Match());
+        }
+    }
+
+    totalMatches() {
+        return this.players.length - 1;
+    }
+
+    generateTournamentBracket() {
+        var totalRounds = this.totalRounds();
+
+        this.matches.forEach( function(match, i, matches) {
+            var round = totalRounds - parseInt(Math.log2(i + 1));
```

Comment:
The indexes make up a complex logic, based on log, +1 increments... This is very error prone.
Have you considered creating matches from previous matches like:
new Match([match1, match2])
It will get the round like match1.round + 1.